pkg/store/copr/region_cache.go

```diff
@@ -166,38 +166,28 @@ func (c *RegionCache) splitKeyRangesByLocation(loc *tikv.KeyLocation, ranges *Ke
 // UnspecifiedLimit means no limit.
 const UnspecifiedLimit = -1
 
-// SplitKeyRangesByLocationsWithBuckets splits the KeyRanges by logical info in the cache.
-// The buckets in the returned LocationKeyRanges are not empty if the region is split by bucket.
-func (c *RegionCache) SplitKeyRangesByLocationsWithBuckets(bo *Backoffer, ranges *KeyRanges, limit int) ([]*LocationKeyRanges, error) {
-	res := make([]*LocationKeyRanges, 0)
-	for ranges.Len() > 0 {
-		if limit != UnspecifiedLimit && len(res) >= limit {
-			break
-		}
-		loc, err := c.LocateKey(bo.TiKVBackoffer(), ranges.At(0).StartKey)
-		if err != nil {
-			return res, derr.ToTiDBErr(err)
-		}
-
-		isBreak := false
-		res, ranges, isBreak = c.splitKeyRangesByLocation(loc, ranges, res)
-		if isBreak {
-			break
-		}
-	}
-
-	return res, nil
-}
-
-// SplitKeyRangesByLocationsWithoutBuckets splits the KeyRanges by logical info in the cache.
+// SplitKeyRangesByLocations splits the KeyRanges by logical info in the cache.
 // The buckets in the returned LocationKeyRanges are empty, regardless of whether the region is split by bucket.
-func (c *RegionCache) SplitKeyRangesByLocationsWithoutBuckets(bo *Backoffer, ranges *KeyRanges, limit int) ([]*LocationKeyRanges, error) {
+func (c *RegionCache) SplitKeyRangesByLocations(bo *Backoffer, ranges *KeyRanges, limit int, needLeader, buckets bool) ([]*LocationKeyRanges, error) {
 	if limit == 0 || ranges.Len() <= 0 {
 		return nil, nil
 	}
-	// Currently, LocationKeyRanges returned by `LocateKeyRange` doesn't contains buckets,
-	// because of https://github.com/tikv/client-go/blob/09ecb550d383c1b048119b586fb5cda658312262/internal/locate/region_cache.go#L1550-L1551.
-	locs, err := c.LocateKeyRange(bo.TiKVBackoffer(), ranges.RefAt(0).StartKey, ranges.RefAt(ranges.Len()-1).EndKey)
+
+	kvRanges := make([]tikv.KeyRange, 0, ranges.Len())
+	for i := 0; i < ranges.Len(); i++ {
+		kvRanges = append(kvRanges, tikv.KeyRange{
+			StartKey: ranges.At(i).StartKey,
+			EndKey:   ranges.At(i).EndKey,
+		})
+	}
+	opts := make([]tikv.BatchLocateKeyRangesOpt, 0, 2)
+	if needLeader {
+		opts = append(opts, tikv.WithNeedRegionHasLeaderPeer())
+	}
+	if buckets {
+		opts = append(opts, tikv.WithNeedBuckets())
+	}
+	locs, err := c.BatchLocateKeyRanges(bo.TiKVBackoffer(), kvRanges, opts...)
```

Comment:
Should the result be returned here directly if limit is `UnspecifiedLimit`  and batch load is used?

Reply:
Types are different, we need cast it from [`KeyLocation`](https://github.com/tikv/client-go/blob/14f3d2b5b785b7458f87d95b4ff24578e6e41184/internal/locate/region_cache.go#L1055) to [`LocationKeyRanges`](https://github.com/pingcap/tidb/blob/1368bf75f11d5406d0f447b3eed1cd625fd826a8/pkg/store/copr/region_cache.go#L64) anyway, in which ranges are splited to the result locations.